tests/end_to_end_tests/managed_packages_tests/conftest.py

```diff
@@ -0,0 +1,409 @@
+                    package_path,
+                    repo_path
+                )
+            subprocess.check_call([
+                "createrepo_c",
+                str(repo_path)
+            ])
+
+    @classmethod
+    def add_command_line_arguments(cls, parser: argparse.ArgumentParser):
+        super(RepoBuilder, cls).add_command_line_arguments(parser)
+
+        parser.add_argument(
+            "--package-type",
+            dest="package_type",
+            required=True
+        )
+        parser.add_argument(
+            "--packages-dir",
+            dest="packages_dir",
+            required=True,
+        )
+
+    @classmethod
+    def handle_command_line_arguments(
+        cls,
+        args,
+    ):
+        super(RepoBuilder, cls).handle_command_line_arguments(args)
+        builder = cls()
+        builder.build(
+            package_type=args.package_type,
+            packages_dir_path=pl.Path(args.packages_dir)
+        )
+
+
+@pytest.fixture(scope="session")
+def repo_dir(package_builder_name, package_builder, request, tmp_path_factory):
+
+    """
+    Fixture with directory where tested packages are stored.
+    Depending on provided tests arguments, it may be a directory with ready repo, or just a directory
+    with packages. In the second case repo is built inplace.
+    """
+    packages_dir = None
+    repo_dir = None
+
+    package_source = request.config.option.packages_source
+
+    if package_source is not None:
+        packages_source_type = request.config.option.packages_source_type
+
+        if packages_source_type == "repo":
+            repo_dir = pl.Path(package_source)
+        elif packages_source_type == "dir":
+            packages_dir = pl.Path(package_source)
+        else:
+            raise Exception(f"Unknown package source type {packages_source_type}")
+
+    if repo_dir:
+        # packages are already in the form of repo.
+        if repo_dir.is_file():
+            with tarfile.open(repo_dir) as tf:
+                repo_dir = tmp_path_factory.mktemp("repo")
+                tf.extractall(repo_dir)
+
+        return repo_dir
+
+    # No repo provided, look for packages to build repo.
+
+    if packages_dir is None:
+        # packages are also not provided, build the now.
+        packages_dir = tmp_path_factory.mktemp("packages")
+        builder_cls = ALL_MANAGED_PACKAGE_BUILDERS[package_builder_name]
+        builder = builder_cls()
+        builder.build_packages()
+        shutil.copytree(
+            builder.packages_output_path,
+            packages_dir,
+            dirs_exist_ok=True
+        )
+
+    # Build repo from packages.
+    repo_builder = RepoBuilder()
+    repo_builder.build(
+        package_type=package_builder.PACKAGE_TYPE,
+        packages_dir_path=packages_dir
+    )
+    repo_dir = repo_builder.output_path
+
+    return repo_dir
+
+
+@pytest.fixture(scope="session")
+def repo_url(package_builder_name, package_builder, repo_dir, tmp_path_factory):
```

Comment:
I was hoping we could keep this simpler - e.g. only install .rpm/.deb packages directly from file, but yeah I do understand the reasoning with testing the complete flow and "installing packages from the package repo".

At least for the stable tests (as we do for the AMI tests today) which run periodically, we should use actual upstream stable packagecloud repo, right? (so we exercise that repo code path and ensure repo is in order)

Reply:
Correct, for the stable tests, as we do now, we just use url of existing repo.

Yeah, maybe I've gone too far by trying to mimic real repo for the dev testing, and it should be enough to do it on only on stable tests. :sweat_smile:

I'll simplify that part.

Reply:
Thanks - I'm always for simplicity and as at least abstractions / indirections as possible to make understanding, troubleshooting, etc. easier :)